File: packages/swift/Sources/SubmitCMS/Modules/Records.swift
import Foundation

/// İçerik kayıtları — v2 şema sisteminin ana modülü (yazma tarafı).
///
/// Ziyaretçiye içerik göstermek için `sdk.delivery` kullanın.
///
/// Gövde şeması:
/// ```swift
/// try await sdk.records.create("blog", payload: [
///     "data": ["baslik": "Merhaba"],   // tipin özel alanları
///     "status": "published",            // draft | published | archived
///     "locale": "tr",                   // sitenin dil listesinde olmalı
///     "categories": [3, 7],
///     "seo": ["meta_title": "..."],
///     "commerce": ["price": 199.90],    // yalnızca ürün tiplerinde
/// ])
/// ```
public struct RecordsModule: Sendable {
    let client: SubmitClient

    /// Kayıtları listeler. Filtre işleçleri: `eq`, `ne`, `gt`, `gte`, `lt`, `lte`, `like`, `in`.
    ///
    /// ```swift
    /// try await sdk.records.list("urun", params: [
    ///     "status": "published",
    ///     "filter": ["price": ["gte": 100]],
    /// ])
    /// ```
    public func list(_ typeCode: String, params: [String: Any] = [:]) async throws -> SubmitResponse<[JSONValue]> {
        try await client.get("/api/schema/records/\(esc(typeCode))", query: params, as: [JSONValue].self)
    }

    public func get(_ typeCode: String, id: Int) async throws -> SubmitResponse<JSONValue> {
        try await client.get("/api/schema/records/\(esc(typeCode))/\(id)")
    }

    /// Yeni kayıt. Plan limiti burada uygulanır — kota dolduysa 403.
    public func create(_ typeCode: String, payload: [String: Any]) async throws -> SubmitResponse<JSONValue> {
        try await client.post("/api/schema/records/\(esc(typeCode))", body: payload)
    }

    /// Kısmi güncelleme — yalnızca gönderdiğiniz alanlar değişir.
    public func update(_ typeCode: String, id: Int, payload: [String: Any]) async throws -> SubmitResponse<JSONValue> {
        try await client.put("/api/schema/records/\(esc(typeCode))/\(id)", body: payload)
    }

    @discardableResult
    /// Kaydı ÇÖP KUTUSUNA taşır (2026-08-13'ten beri iki aşamalı).
    ///
    /// Varsayılan silme kalıcı DEĞİLDİR: kayıt siteden düşer, 30 gün içinde
    /// `restoreFromTrash` ile geri alınabilir, süre dolunca sunucudaki
    /// `records:prune` görevi kalıcı siler.
    ///
    /// `force: true` kalıcı siler ve YALNIZ şirket yöneticisinde çalışır;
    /// yetkisi olmayan çağrı 403 döner (sessizce çöpe düşmez).
    public func delete(_ typeCode: String, id: Int, force: Bool = false) async throws -> SubmitResponse<JSONValue> {
        let path = "/api/schema/records/\(esc(typeCode))/\(id)"

        return try await client.delete(force ? path + "?force=1" : path)
    }

    /// Çöp kutusu: silinmiş ama henüz kalıcı silinmemiş kayıtlar.
    ///
    /// `meta.can_purge` çağıran kullanıcının kalıcı silme yetkisi olup
    /// olmadığını söyler; arayüz "Kalıcı sil" düğmesini buna bakarak çizmelidir.
    public func trash(_ typeCode: String, params: [String: Any] = [:]) async throws -> SubmitResponse<[JSONValue]> {
        try await client.get("/api/schema/records/\(esc(typeCode))/trash", query: params, as: [JSONValue].self)
    }

    /// Kaydı çöp kutusundan geri getirir.
    ///
    /// Adres çakışması sunucuda çözülür: kayıt çöpteyken slug'ı başkasına
    /// verilmiş olabilir; geri gelen kayıt CANLI sayfanın adresini almaz,
    /// kendine yeni adres alır. Filtre indeksi de bu sırada yeniden kurulur.
    public func restoreFromTrash(_ typeCode: String, id: Int) async throws -> SubmitResponse<JSONValue> {
        try await client.post("/api/schema/records/\(esc(typeCode))/\(id)/restore", body: [:])
    }

    public func analytics(_ typeCode: String, id: Int) async throws -> SubmitResponse<JSONValue> {
        try await client.get("/api/schema/records/\(esc(typeCode))/\(id)/analytics")
    }

    public func revisions(_ typeCode: String, id: Int) async throws -> SubmitResponse<[JSONValue]> {
        try await client.get("/api/schema/records/\(esc(typeCode))/\(id)/revisions", as: [JSONValue].self)
    }

    public func revision(_ typeCode: String, id: Int, version: Int) async throws -> SubmitResponse<JSONValue> {
        try await client.get("/api/schema/records/\(esc(typeCode))/\(id)/revisions/\(version)")
    }

    /// Kaydı o sürüme döndürür; mevcut hâl önce anlık görüntüye alınır.
    public func restoreRevision(_ typeCode: String, id: Int, version: Int) async throws -> SubmitResponse<JSONValue> {
        try await client.post("/api/schema/records/\(esc(typeCode))/\(id)/revisions/\(version)/restore")
    }

    public func gallery(_ typeCode: String, id: Int, field: String) async throws -> SubmitResponse<[JSONValue]> {
        try await client.get("/api/schema/records/\(esc(typeCode))/\(id)/gallery/\(esc(field))", as: [JSONValue].self)
    }

    public func addGalleryImage(_ typeCode: String, id: Int, field: String, payload: [String: Any]) async throws -> SubmitResponse<JSONValue> {
        try await client.post("/api/schema/records/\(esc(typeCode))/\(id)/gallery/\(esc(field))", body: payload)
    }

    /// Görsel sırasını topluca günceller — verilen id dizisi yeni sıradır.
    public func reorderGallery(_ typeCode: String, id: Int, field: String, order: [Int]) async throws -> SubmitResponse<JSONValue> {
        try await client.put("/api/schema/records/\(esc(typeCode))/\(id)/gallery/\(esc(field))/order", body: ["order": order])
    }

    public func updateGalleryImage(_ typeCode: String, id: Int, field: String, imageId: Int, payload: [String: Any]) async throws -> SubmitResponse<JSONValue> {
        try await client.put("/api/schema/records/\(esc(typeCode))/\(id)/gallery/\(esc(field))/\(imageId)", body: payload)
    }

    @discardableResult
    public func removeGalleryImage(_ typeCode: String, id: Int, field: String, imageId: Int) async throws -> SubmitResponse<JSONValue> {
        try await client.delete("/api/schema/records/\(esc(typeCode))/\(id)/gallery/\(esc(field))/\(imageId)")
    }

    /// Metni iyileştirir. AI kredisi harcar — bakiye yetmezse 402.
    public func aiImprove(_ typeCode: String, id: Int, payload: [String: Any] = [:]) async throws -> SubmitResponse<JSONValue> {
        try await client.post("/api/schema/records/\(esc(typeCode))/\(id)/ai/improve", body: payload)
    }

    public func aiSeo(_ typeCode: String, id: Int, payload: [String: Any] = [:]) async throws -> SubmitResponse<JSONValue> {
        try await client.post("/api/schema/records/\(esc(typeCode))/\(id)/ai/seo", body: payload)
    }

    /// Kaydı hedef dile çevirip kardeş kayıt olarak bağlar.
    public func aiTranslate(_ typeCode: String, id: Int, locale: String) async throws -> SubmitResponse<JSONValue> {
        try await client.post("/api/schema/records/\(esc(typeCode))/\(id)/ai/translate", body: ["locale": locale])
    }
}

File: packages/kotlin/bin/main/com/submitcms/sdk/modules/Records.kt
package com.submitcms.sdk.modules

import com.submitcms.sdk.SubmitClient
import kotlinx.serialization.json.JsonObject

/**
 * İçerik kayıtları — v2 şema sisteminin ana modülü (yazma tarafı).
 *
 * Ziyaretçiye içerik göstermek için [Delivery] kullanın.
 *
 * ```kotlin
 * sdk.records.create("blog", mapOf(
 *     "data" to mapOf("baslik" to "Merhaba"),  // tipin özel alanları
 *     "status" to "published",                  // draft | published | archived
 *     "locale" to "tr",                         // sitenin dil listesinde olmalı
 *     "categories" to listOf(3, 7),
 *     "seo" to mapOf("meta_title" to "..."),
 *     "commerce" to mapOf("price" to 199.90),   // yalnızca ürün tiplerinde
 * ))
 * ```
 */
class Records(private val client: SubmitClient) {

    /**
     * Kayıtları listeler. Filtre işleçleri: `eq`, `ne`, `gt`, `gte`, `lt`, `lte`, `like`, `in`.
     *
     * ```kotlin
     * sdk.records.list("urun", mapOf(
     *     "status" to "published",
     *     "filter" to mapOf("price" to mapOf("gte" to 100)),
     * ))
     * ```
     */
    suspend fun list(typeCode: String, params: Map<String, Any?> = emptyMap()): JsonObject =
        client.get("/api/schema/records/${esc(typeCode)}", params)

    suspend fun get(typeCode: String, id: Int): JsonObject =
        client.get("/api/schema/records/${esc(typeCode)}/$id")

    /** Yeni kayıt. Plan limiti burada uygulanır — kota dolduysa 403. */
    suspend fun create(typeCode: String, payload: Map<String, Any?>): JsonObject =
        client.post("/api/schema/records/${esc(typeCode)}", payload)

    /** Kısmi güncelleme — yalnızca gönderdiğiniz alanlar değişir. */
    suspend fun update(typeCode: String, id: Int, payload: Map<String, Any?>): JsonObject =
        client.put("/api/schema/records/${esc(typeCode)}/$id", payload)

    suspend fun delete(typeCode: String, id: Int): JsonObject =
        client.delete("/api/schema/records/${esc(typeCode)}/$id")

    suspend fun analytics(typeCode: String, id: Int): JsonObject =
        client.get("/api/schema/records/${esc(typeCode)}/$id/analytics")

    suspend fun revisions(typeCode: String, id: Int): JsonObject =
        client.get("/api/schema/records/${esc(typeCode)}/$id/revisions")

    suspend fun revision(typeCode: String, id: Int, version: Int): JsonObject =
        client.get("/api/schema/records/${esc(typeCode)}/$id/revisions/$version")

    /** Kaydı o sürüme döndürür; mevcut hâl önce anlık görüntüye alınır. */
    suspend fun restoreRevision(typeCode: String, id: Int, version: Int): JsonObject =
        client.post("/api/schema/records/${esc(typeCode)}/$id/revisions/$version/restore")

    suspend fun gallery(typeCode: String, id: Int, field: String): JsonObject =
        client.get("/api/schema/records/${esc(typeCode)}/$id/gallery/${esc(field)}")

    suspend fun addGalleryImage(typeCode: String, id: Int, field: String, payload: Map<String, Any?>): JsonObject =
        client.post("/api/schema/records/${esc(typeCode)}/$id/gallery/${esc(field)}", payload)

    /** Görsel sırasını topluca günceller — verilen id dizisi yeni sıradır. */
    suspend fun reorderGallery(typeCode: String, id: Int, field: String, order: List<Int>): JsonObject =
        client.put("/api/schema/records/${esc(typeCode)}/$id/gallery/${esc(field)}/order", mapOf("order" to order))

    suspend fun updateGalleryImage(
        typeCode: String,
        id: Int,
        field: String,
        imageId: Int,
        payload: Map<String, Any?>,
    ): JsonObject =
        client.put("/api/schema/records/${esc(typeCode)}/$id/gallery/${esc(field)}/$imageId", payload)

    suspend fun removeGalleryImage(typeCode: String, id: Int, field: String, imageId: Int): JsonObject =
        client.delete("/api/schema/records/${esc(typeCode)}/$id/gallery/${esc(field)}/$imageId")

    /** Metni iyileştirir. AI kredisi harcar — bakiye yetmezse 402. */
    suspend fun aiImprove(typeCode: String, id: Int, payload: Map<String, Any?> = emptyMap()): JsonObject =
        client.post("/api/schema/records/${esc(typeCode)}/$id/ai/improve", payload)

    suspend fun aiSeo(typeCode: String, id: Int, payload: Map<String, Any?> = emptyMap()): JsonObject =
        client.post("/api/schema/records/${esc(typeCode)}/$id/ai/seo", payload)

    /** Kaydı hedef dile çevirip kardeş kayıt olarak bağlar. */
    suspend fun aiTranslate(typeCode: String, id: Int, locale: String): JsonObject =
        client.post("/api/schema/records/${esc(typeCode)}/$id/ai/translate", mapOf("locale" to locale))
}
